File: MNN/cpp/src/mnn_run.cpp
#include "mnn_run.h"


/**
 * @brief Construct a new MNNModel::MNNModel object
 * 
 * @param mnnPath local path to *.mnn model file
 * @param mode inference mode, fp32/fp16/half
 * @param device inference device and driver, cpu/opencl/opengl/vulkan/metal
 */
MNNModel::MNNModel(const string& mnnPath, const string& mode, const string& device):
    mnnPath(mnnPath), mode(mode), device(device) {

    this->init();
}

/**
 * @brief init mnn network
 * 
 */
void MNNModel::init() {
    struct stat buffer;
    if (stat(this->mnnPath.c_str(), &buffer) != 0) {
        cout << "model file: " << this->mnnPath << " not exists!" << endl; 
        assert(0);
    }

    // build network
    this->net = Interpreter::createFromFile(this->mnnPath.c_str());

    // build config
    ScheduleConfig config;

    // set cpu thread used
    config.numThread = this->numThread;

    // set host device
    if (this->device == "cpu") {
        config.type = static_cast<MNNForwardType>(MNN_FORWARD_CPU);
    } else if (this->device == "opencl") {
        config.type = static_cast<MNNForwardType>(MNN_FORWARD_OPENCL);
    } else if (this->device == "opengl") {
        config.type = static_cast<MNNForwardType>(MNN_FORWARD_OPENGL);
    } else if (this->device == "vulkan") {
        config.type = static_cast<MNNForwardType>(MNN_FORWARD_VULKAN);
    } else if (this->device == "metal") {
        config.type = static_cast<MNNForwardType>(MNN_FORWARD_METAL);
    } else {
        cout << "Un-support device: " << this->device << endl; 
        assert(0);
    }

    // set precision
    BackendConfig backendConfig;
    if (this->mode == "fp16") {
        backendConfig.precision = static_cast<BackendConfig::PrecisionMode>(BackendConfig::Precision_Low);
    } else if (this->mode == "half") {
        backendConfig.precision = static_cast<BackendConfig::PrecisionMode>(BackendConfig::Precision_Normal);
    } else if (this->mode == "fp32") {
        backendConfig.precision = static_cast<BackendConfig::PrecisionMode>(BackendConfig::Precision_High);
    } else {
        cout << "Un-support mode: " << this->mode << endl; 
        assert(0);
    }

    // set power use
    backendConfig.power = static_cast<BackendConfig::PowerMode>(BackendConfig::Power_Normal);
    // set memory use
    backendConfig.memory = static_cast<BackendConfig::MemoryMode>(BackendConfig::Memory_Normal);

    config.backendConfig = &backendConfig;

    // build session use config
    this->session = this->net->createSession(config);

    // get input and output node of network
    this->modelInputTensor = this->net->getSessionInput(this->session, NULL);
    this->modelOutputTensor = this->net->getSessionOutput(this->session, NULL);
    this->hostInputTensor = new Tensor(this->modelInputTensor, Tensor::CAFFE);
    this->hostOutputTensor = new Tensor(this->modelOutputTensor, Tensor::CAFFE);
}

/**
 * @brief inference on mnn
 * 
 * @param image opencv image
 * @param output output float data
 */
void MNNModel::forawrd(const cv::Mat& image, vector<float>& output) {
    memcpy(this->hostInputTensor->host<float>(), image.data, this->hostInputTensor->size());

    this->modelInputTensor->copyFromHostTensor(this->hostInputTensor);
    net->runSession(session);
    this->modelOutputTensor->copyToHostTensor(this->hostOutputTensor);

    output.resize(this->hostOutputTensor->elementSize());
    memcpy(output.data(), this->hostOutputTensor->host<float>(), this->hostOutputTensor->size());
}

/**
 * @brief Destroy the MNNModel::MNNModel object
 * 
 */
MNNModel::~MNNModel() {
    delete this->hostInputTensor;
    delete this->hostOutputTensor;

    if (this->net != nullptr)
        delete this->net;
}

File: MNN/python/mnn_run.py
import os
import MNN
import numpy as np


class MNNModel:
    def __init__(self, model_path, device="CPU"):
        """
        :param model_path: local path of mnn file.
        :param device: run device, CPU only for pypi.
        """
        if not os.path.exists(model_path):
            raise Exception("model file {} not exist!".format(model_path))

        self.device = device
        self._get_session(model_path)

    def _get_session(self, model_path):
        """build run session
        :param model_path: local path of mnn file.
        """
        self.interpreter = MNN.Interpreter(model_path)

        # config
        config = {}
        config['backend'] = self.device
        config['precision'] = 'low'
        config['numThread'] = 1

        # create session
        self.session = self.interpreter.createSession(config)
        self.interpreter.resizeSession(self.session)

        # input and output nodes of model
        self.input_tensors = self.interpreter.getSessionInputAll(self.session)
        self.output_tensors = self.interpreter.getSessionOutputAll(self.session)

        for i, (k, v) in enumerate(self.input_tensors.items()):
            print("input index: {}, name: {}, shape: {}".format(i, k, v.getShape()))
        for i, (k, v) in enumerate(self.output_tensors.items()):
            print("output index: {}, name: {}, shape: {}".format(i, k, v.getShape()))

    def forward(self, image_tensors):
        """do infernece
        :param image_tensors: list, inputs of model.
        return outputs: list, outputs of model.
        """
        outputs = []

        # copy from host to device
        for i, (k, v) in enumerate(self.input_tensors.items()):
            image = image_tensors[i].astype(np.float32)
            tmp_input = MNN.Tensor(v.getShape(), MNN.Halide_Type_Float, image, MNN.Tensor_DimensionType_Caffe)
            v.copyFrom(tmp_input)

        # run
        self.interpreter.runSession(self.session)

        # copy from device to host
        for i, (k, v) in enumerate(self.output_tensors.items()):
            tmp_output = MNN.Tensor(v.getShape(), MNN.Halide_Type_Float, np.ones(v.getShape()).astype(np.float32), MNN.Tensor_DimensionType_Caffe)
            v.copyToHostTensor(tmp_output)
            outputs.append(np.array(tmp_output.getData()))

        return outputs
